Index: default/eng_stringtable.txt
===================================================================
--- default/eng_stringtable.txt	(revision 4590)
+++ default/eng_stringtable.txt	(working copy)
@@ -3173,9 +3173,11 @@
 # Ship Slot Types
 
 SL_EXTERNAL
-External
+External Slot
 SL_INTERNAL
-Internal
+Internal Slot
+SL_TOOLTIP_DESC
+Empty
 
 
 # Ship Part Classes
@@ -6330,7 +6332,6 @@
 SP_DISTORTION_MODULATOR_DESC
 Distinguishes natural geometric irregularities from enemy ships by manipulating space-time distortions caused by all objects in the vicinity.  All objects within 0 uus have max stealth reduced by 20.  The effect is non-stacking.
 
-
 ##############
 # Ship Hulls #
 ##############
Index: UI/DesignWnd.cpp
===================================================================
--- UI/DesignWnd.cpp	(revision 4590)
+++ UI/DesignWnd.cpp	(working copy)
@@ -1454,6 +1454,7 @@
     m_background(0)
 {
     SetDragDropDataType("");
+
 }
 
 SlotControl::SlotControl(double x, double y, ShipSlotType slot_type) :
@@ -1471,6 +1472,13 @@
                                          GG::GRAPHIC_FITGRAPHIC | GG::GRAPHIC_PROPSCALE);
     m_background->Show();
     AttachChild(m_background);
+
+    SetBrowseModeTime(GetOptionsDB().Get<int>("UI.tooltip-delay"));
+
+    // set up empty slot tool tip
+    SetBrowseInfoWnd(boost::shared_ptr<GG::BrowseInfoWnd>(
+        new IconTextBrowseWnd(SlotBackgroundTexture(m_slot_type), UserString(((slot_type == SL_EXTERNAL) ? "SL_EXTERNAL" : "SL_INTERNAL")), UserString("SL_TOOLTIP_DESC"))));
+
 }
 
 void SlotControl::DropsAcceptable(DropsAcceptableIter first, DropsAcceptableIter last, const GG::Pt& pt) const {
@@ -1599,6 +1607,11 @@
         // double click clears slot
         GG::Connect(m_part_control->DoubleClickedSignal,
                     boost::bind(&SlotControl::EmitNullSlotContentsAlteredSignal, this));
+        SetBrowseModeTime(GetOptionsDB().Get<int>("UI.tooltip-delay"));
+
+        // set part occupying slot's tool tip to say slot type as well
+        m_part_control->SetBrowseInfoWnd(boost::shared_ptr<GG::BrowseInfoWnd>(
+            new IconTextBrowseWnd(ClientUI::PartTexture(part_type->Name()), UserString(part_type->Name()) + " (" + UserString(((m_slot_type == SL_EXTERNAL) ? "SL_EXTERNAL" : "SL_INTERNAL")) + ")" ,UserString(part_type->Description()))));
     }
 }
 
